mantidproject/mantid
# Allow switching between Python 2/3 without cleaning CMake configuration

**Description of work.**

Currently switching between Python 2 and Python 3 requires removing the `CMakeCache.txt` file. These changes introduce checks on the old Python configuration to ensure it matches with the requested Python version. If not they are cleared out and rediscovered. A new `WITH_PYTHON3` option has been introduced globally to control whether we build against Python 2/3 to allow the default to be switched to 3 more easily.



**To test:**

This needs to be tested using an existing build tree.

* Once the branch is checked out run cmake as normal without any extra options. It should print that Python 2 is used. Check the build uses python2 and boost_python in either the MSVC .vcxproj or the ninja files 
* Once configuration is complete run cmake again but with the `WITH_PYTHON3` option set to ON. It should print out the Python 3 version and a message `Python version has changed. Clearing previous Python configuration.`. Check Check the build uses python3 and boost_python3 in either the MSVC .vcxproj or the ninja files.

*There is no associated issue.*

*This does not require release notes* because **it is an internal change.**

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
